server/aesdsocket.c

```diff
@@ -0,0 +1,256 @@
+	if(daemon == 1 ) {
+		pid_t pid;
+		printf("Run as daemon\n\r");
+		
+		signal(SIGCHLD, SIG_IGN);
+	        signal(SIGHUP, SIG_IGN);
+	        
+		pid = fork();
+		if(pid == -1)
+			return -1;
+		else if(pid != 0)
+			exit(EXIT_SUCCESS);
+
+		umask(0);
+
+		if(setsid() == -1)
+			return -1;
+
+		if(chdir("/") == -1)
+			return -1;
+			
+		close(STDIN_FILENO);
+		close(STDOUT_FILENO);
+		close(STDERR_FILENO);
+		
+		//redirect files to /dev/null
+		open("/dev/null", O_RDWR);
+		dup(0);
+		dup(0);
+	}
+
+	//listen for connections
+	rc = listen(sockfd, MAX_CONNECTS);
+	if(rc<0) {
+		perror("Server listen");
+		return -1;
+	}
+	
+	//open /var/tmp/aesdsocketdata file
+	filefd = open(output_file, O_RDWR | O_CREAT, 0644);
+	if(filefd < 0) {
+		perror("Open file");
+		return -1;
+	}
+
+	while(1) {
+
+		addr_size = sizeof(new_addr);
+		
+		//accept socket connections
+		new_sockfd = accept(sockfd, (struct sockaddr*)&new_addr, &addr_size);
+		if(new_sockfd<0) {
+			perror("Server accept");
+			return -1;
+		}
+		
+		inet_ntop(AF_INET, &ipAddr, str, INET_ADDRSTRLEN);
+		syslog(LOG_INFO, "Accepted connection from %s", str);
+
+		long write_byte=0, recv_byte=0, send_byte=0, read_byte=0;
+		char *read_buffer = NULL;
+		char *write_buffer = NULL;
+		long rdbuff_size = 0;
+		long malloc_size = BUFFER_MAX;
+		int break_loop=0;
+		
+		read_buffer = (char *)malloc(sizeof(char)*BUFFER_MAX);
+		if(read_buffer == NULL) {
+			perror("Read Malloc failed");
+			return -1;
+		}
+
+		do {
+			//receive data over socket and store it in buffer
+			recv_byte = recv(new_sockfd, buffer, sizeof(buffer), 0);
+			
+			//check for new-line character
+			if(!recv_byte || (strchr(buffer, '\n')!=NULL))
+				break_loop=1;
+
+			//check if allocated buffer size is sufficient
+			if((malloc_size - rdbuff_size) < recv_byte) {
+
+				malloc_size += recv_byte;
+				//realloc required size
+				read_buffer = (char *)realloc(read_buffer, sizeof(char)*malloc_size);
+			}
+			
+			//copy received buffer in another read_buffer
+			memcpy(&read_buffer[rdbuff_size], buffer, recv_byte);
+			rdbuff_size += recv_byte;
+			
+		} while(break_loop != 1);
+		
+		//write from read_buffer in output file
+		write_byte = write(filefd, read_buffer, rdbuff_size);
+		if(write_byte != rdbuff_size) {
+			perror("File write");
+			return -1;
+		}
+		
+		//seek the start of the file
+		lseek(filefd, 0, SEEK_SET);
+		
+		//send contents writen in output file to client line by line
+		wrbuff_size += rdbuff_size;
+		write_buffer = (char *)malloc(sizeof(char)*wrbuff_size);
+		if(write_buffer == NULL) {
+			perror("Write malloc failed");
+			return -1;
+		}
+			
+		//store contents os output file in write_buffer
+		read_byte = read(filefd, write_buffer, wrbuff_size);
+		if(read_byte != wrbuff_size) {
+			perror("File read");
+			return -1;
+		}
+		
+		//send contents of write_buffer to client
+		send_byte = send(new_sockfd, write_buffer, wrbuff_size, 0);
+		if(send_byte != wrbuff_size) {
+			perror("Socket send");
```

Comment:
@saloni1307 missing cleanup in the error paths here.  Consider setting a variable for return status and exiting through the normal path with associated cleanup in one place to avoid code duplication.